crates/snforge-scarb-plugin/src/utils.rs

```diff
@@ -1,30 +1,51 @@
 use cairo_lang_macro::{Diagnostic, Severity};
 use indoc::formatdoc;
 
-fn higher_severity(a: Severity, b: Severity) -> Severity {
+pub fn higher_severity(a: Severity, b: Severity) -> Severity {
     match (a, b) {
         (Severity::Warning, Severity::Warning) => Severity::Warning,
         _ => Severity::Error,
     }
 }
+pub fn format_error_message(variants: &[Diagnostic]) -> String {
+    let formatted_variants: Vec<String> = variants
+        .iter()
+        .map(|variant| format!("- variant: {}", variant.message))
+        .collect();
 
-pub fn branch(
-    left: Result<String, Diagnostic>,
-    right: impl Fn() -> Result<String, Diagnostic>,
-) -> Result<String, Diagnostic> {
-    left.or_else(|error| {
-        right().map_err(|next_error| Diagnostic {
-            severity: higher_severity(error.severity, next_error.severity),
-            message: formatdoc!(
-                "
-                    Both options failed
-                    First variant: {}
-                    Second variant: {}
-                    Resolve at least one of them
-                ",
-                error.message,
-                next_error.message
-            ),
-        })
-    })
+    formatdoc! {"
+        All options failed
+        {}
+        Resolve at least one of them
+    ", formatted_variants.join("\n")}
+}
+
+#[macro_export]
+macro_rules! branch {
```

Comment:
I don't know if this is a macro. It would be more readable as a function.

```rs
pub fn branch(results: Vec<Result<String, Diagnostic>>) -> Result<String, Diagnostic> {
    let mut messages: Vec<Diagnostic> = Vec::new();

    for result in results {
        match result {
            Ok(val) => return Ok(val),
            Err(err) => messages.push(err),
        }
    }

    Err(Diagnostic {
        message: format_error_message(&messages),
        severity: messages.iter().fold(Severity::Warning, |acc, diagnostic| {
            higher_severity(acc, diagnostic.severity)
        }),
    })
}
```

Reply:
Macro is lazy

Reply:
Still can be done with function though 

```rs
pub fn branch(results: Vec<Box<dyn Fn() -> Result<String, Diagnostic>>>) -> Result<String, Diagnostic> {
    let mut messages: Vec<Diagnostic> = Vec::new();

    for result in results {
        match result() {
            Ok(val) => return Ok(val),
            Err(err) => messages.push(err),
        }
    }

    Err(Diagnostic {
        message: format_error_message(&messages),
        severity: messages.iter().fold(Severity::Warning, |acc, diagnostic| {
            higher_severity(acc, diagnostic.severity)
        }),
    })
}

```

Reply:
I'll stay with the macro because it has better readability

Reply:
No problem 😄. But I think you are mistaking verbosity for readability."